### DISCUSSION TITLE
Can't use `document()` inside a function

### DISCUSSION THREAD

**Discussion by Andrew15-5** [typst/typst#2127] [General]
### Description

This code doesn't do anything:

```typ
#let set-title(title) = {
  set document(title: title)
}
#set-title("custom title")
```

But it should've changed the title metadata/entry in the PDF file. Same way, I can't change the author of the document.

### Operating system

Linux

### Typst version

- [X] I am using the latest version of Typst

---

Idea discussion: #2153.

**Comment by Myriad-Dreamin** [comment]
If I understand correctly, it works as expected, since the set rules only affect its belonging scope.

```
#let set-title(title) = { // scope inner
  set document(title: title)
}

// scope outer

#set-title("custom title")
```

**Comment by LuxxxLucy** [comment]
Relevant #1913 

Setting a variable inside the scope of the function does not change the outer variable. And this is what is expected. Maybe you should follow conventional way of using #set_title or  #show doc => template(title: "custom title", doc) in the beginning of the document instead of a function.

**Comment by Andrew15-5** [comment]
I know about the scope and everything like that, but aren't there any other way besides using `#show: template.with(title: "title")`?

Maybe it's just functional programming got a hold of me.

**Comment by Andrew15-5** [comment]
My main problem is that I have a template and a separate function for a title page with a dictionary of values:

```typ
#import "@local/package:0.0.1": template, title-page
#show: template
#title-page((
  title: "title",
  a: 1,
  b: "c",
  d: (3, 4, 5),
))
```

Right now, I have to `set document` between `show` and `title-page`, because it's a global scope. Setting `document` inside `title-page` isn't working. I can do this in `template`. But this means, that I have to do:

```typ
#let title = "title"
#show: template.with(title: title)
#title-page((
  title: title,
))
```

Which I fill is a bit too verbose than just passing the title one to `title-page` and setting the `document` there.
To fix this, I would have to convert `title-page` function to a template that is used in the show rule, which also doesn't feel right.

I don't know, I just want the code to be elegant and minimalistic.

**Comment by Andrew15-5** [comment]
Maybe I want to compose a title from some values that are passed to the `title-page`. This would be so simple: just combine values inside `title-page` and set the title also there. Instead, I have to compose the title in the main file (unwanted mess at the beginning of the main file!). It seems that then I just have to pass all the needed arguments to the `template` and compose the title there, instead of in the `title-page`.

In other words, if making the title makes sense in a function (and therefore setting the title there), because it has all the needed information (and there is literally "title" in `title-page`), you can't. You have to tear off the logic from the function and put it inside the template (main template or a separate one).

**Comment by LuxxxLucy** [comment]
Have you consider passing doc as a argument? 


#let template(title: none, doc) = {
      set document(title: title)
      // do something
      doc
}


#let title-page(title: none, doc) = {
      // do something else. Modify title here
      set document(title: title)
      // do something else
      doc
}

#show doc => template(title: “first title”, doc)

#show doc => title-page(title: “second title”, doc)




I did some quick test and this seems to be working.

**Comment by Andrew15-5** [comment]
> ```typ
> #show doc => template(title: “first title”, doc)
> 
> #show doc => title-page(title: “second title”, doc)
> ```

I already mentioned this:

> To fix this, I would have to convert `title-page` function to a template that is used in the show rule, which also doesn't feel right.

And the title is the same, so I would have to add another line with a variable, which I also already mentioned:

> ```typ
> #let title = "title"
> #show: template.with(title: title)
> #title-page((
>   title: title,
> ))
> ```

**Comment by LuxxxLucy** [comment]
Sorry I cannot comprehend.

Perhaps make a wrapper to get the template and title-page into another (bigger) template and then calling it?

**Reply by Andrew15-5** [reply]
It can be another workaround, but `template` and `title-page` live in different files, because they have nothing (or almost nothing) in common. Your approach will require either making another file that will have the bigger wrapper, or join the code from both files. Either way, it will only complicate the code itself.

Since `title-page` will set the title/first page, it does make sense that it could also set the title of the document itself. But this is unfortunately not possible (yet) in Typst.

**Reply by LuxxxLucy** [reply]
From my understanding, you are correct. 

Because what you want is a function call with side effects. It would not be compatible with the general design of typst, where memorization of function call is a core feature; side effects will break it. 

So it is not something that is **not supported yet**. It is a deliberate design for not supporting it.

**Reply by Andrew15-5** [reply]
I've opened a related discussion here: #2153.
